File: test/lims/lims/client/src/api/services/equipmentService.ts
import apiClient from '../axiosConfig';
import { API_ENDPOINTS } from '../constants';
import { Equipment, EquipmentType, EquipmentStatus } from '../../types/equipment';

class EquipmentService {
    /**
     * Get all equipment with pagination and filtering
     */
    async getAllEquipment(params: {
        page?: number;
        limit?: number;
        type?: string;
        status?: string;
        location?: string;
        manufacturer?: string;
        search?: string;
        sortBy?: string;
        sortOrder?: 'asc' | 'desc';
    } = {}): Promise<{ 
        data: Equipment[],
        pagination: {
            currentPage: number;
            totalPages: number;
            totalItems: number;
            itemsPerPage: number;
            hasMore: boolean;
        } 
    }> {
        const response = await apiClient.get<{
            data: Equipment[],
            pagination: {
                currentPage: number;
                totalPages: number;
                totalItems: number;
                itemsPerPage: number;
                hasMore: boolean;
            },
            status: number,
            success: boolean
        }>(
            API_ENDPOINTS.METADATA.EQUIPMENT,
            { params }
        );
        return {
            data: response.data.data,
            pagination: response.data.pagination
        };
    }

    /**
     * Create new equipment
     */
    async createEquipment(data: Partial<Equipment>): Promise<Equipment> {
        const response = await apiClient.post<{ data: Equipment, status: number, success: boolean }>(
            API_ENDPOINTS.METADATA.EQUIPMENT,
            data
        );
        return response.data.data;
    }

    /**
     * Update equipment
     */
    async updateEquipment(id: string, updates: Partial<Equipment>): Promise<Equipment> {
        const response = await apiClient.put<{ data: Equipment, status: number, success: boolean }>(
            `${API_ENDPOINTS.METADATA.EQUIPMENT}/${id}`,
            updates
        );
        return response.data.data;
    }

    /**
     * Delete equipment
     */
    async deleteEquipment(id: string): Promise<void> {
        await apiClient.delete(`${API_ENDPOINTS.METADATA.EQUIPMENT}/${id}`);
    }

    /**
     * Get equipment types
     */
    async getEquipmentTypes(): Promise<EquipmentType[]> {
        const response = await apiClient.get<{ data: EquipmentType[], status: number, success: boolean }>(
            API_ENDPOINTS.METADATA.EQUIPMENT_TYPES
        );
        return response.data.data;
    }

    /**
     * Get equipment statuses
     */
    async getEquipmentStatuses(): Promise<EquipmentStatus[]> {
        const response = await apiClient.get<{ data: EquipmentStatus[], status: number, success: boolean }>(
            API_ENDPOINTS.METADATA.EQUIPMENT_STATUSES
        );
        return response.data.data;
    }

    /**
     * Add maintenance record
     */
    async addMaintenanceRecord(
        equipmentId: string,
        record: {
            date: string;
            performedBy: string;
            task: string;
            status: string;
            notes?: string;
        }
    ): Promise<Equipment> {
        const response = await apiClient.post<{ data: Equipment, status: number, success: boolean }>(
            `${API_ENDPOINTS.METADATA.EQUIPMENT}/${equipmentId}/maintenance`,
            record
        );
        return response.data.data;
    }

    /**
     * Add note to equipment
     */
    async addNote(equipmentId: string, note: { content: string, user: string }): Promise<Equipment> {
        const response = await apiClient.post<{ data: Equipment, status: number, success: boolean }>(
            `${API_ENDPOINTS.METADATA.EQUIPMENT}/${equipmentId}/notes`,
            note
        );
        return response.data.data;
    }

    /**
     * Add attachment to equipment
     */
    async addAttachment(equipmentId: string, attachment: FormData): Promise<Equipment> {
        const response = await apiClient.post<{ data: Equipment, status: number, success: boolean }>(
            `${API_ENDPOINTS.METADATA.EQUIPMENT}/${equipmentId}/attachments`,
            attachment,
            {
                headers: {
                    'Content-Type': 'multipart/form-data',
                },
            }
        );
        return response.data.data;
    }
}

export default new EquipmentService();

File: server/app/services/equipment_service.py
"""
Equipment/Instrument service layer handling business logic
"""
from typing import List, Optional, Dict, Any
from datetime import datetime
from sqlalchemy.orm import Session
from sqlalchemy import or_
from fastapi import HTTPException, UploadFile
from app.db.models.instrument import (
    Instrument, InstrumentCalibration,
    InstrumentMaintenanceLog, Note
)
from app.api.schemas.equipment import (
    InstrumentCreate, InstrumentUpdate, CalibrationCreate,
    MaintenanceCreate, NoteCreate
)
from app.core.config import settings
from app.core.constants.equipment import (
    EquipmentType, EquipmentStatus,
    MaintenanceStatus, CalibrationStatus
)
from app.core.utils.file_handling import save_upload_file

class EquipmentService:
    @staticmethod
    def get_all_equipment(
        db: Session,
        skip: int = 0,
        limit: int = 10,
        filters: Optional[Dict[str, Any]] = None,
        search: Optional[str] = None
    ) -> Dict[str, Any]:
        """Get all equipment with pagination and filtering"""
        query = db.query(Instrument)

        # Apply filters
        if filters:
            if filters.get("type"):
                query = query.filter(Instrument.instrument_type == filters["type"])
            if filters.get("status"):
                query = query.filter(Instrument.status == filters["status"])
            if filters.get("location_id"):
                query = query.filter(Instrument.location_id == filters["location_id"])
            if filters.get("manufacturer"):
                query = query.filter(Instrument.manufacturer == filters["manufacturer"])

        # Apply search
        if search:
            search_filter = or_(
                Instrument.name.ilike(f"%{search}%"),
                Instrument.description.ilike(f"%{search}%"),
                Instrument.serial_number.ilike(f"%{search}%")
            )
            query = query.filter(search_filter)

        total = query.count()
        items = query.offset(skip).limit(limit).all()

        return {
            "items": items,
            "total": total,
            "page": skip // limit + 1,
            "size": limit,
            "pages": (total + limit - 1) // limit
        }

    @staticmethod
    def get_equipment_by_id(db: Session, equipment_id: int) -> Instrument:
        """Get equipment by ID"""
        equipment = db.query(Instrument).filter(Instrument.id == equipment_id).first()
        if not equipment:
            raise HTTPException(status_code=404, detail="Equipment not found")
        return equipment

    @staticmethod
    def create_equipment(db: Session, equipment: InstrumentCreate) -> Instrument:
        """Create new equipment"""
        # Check for duplicate serial number
        existing = db.query(Instrument).filter(
            Instrument.serial_number == equipment.serial_number
        ).first()
        if existing:
            raise HTTPException(
                status_code=400,
                detail="Equipment with this serial number already exists"
            )

        db_equipment = Instrument(**equipment.dict())
        db.add(db_equipment)
        db.commit()
        db.refresh(db_equipment)
        return db_equipment

    @staticmethod
    def update_equipment(
        db: Session,
        equipment_id: int,
        updates: InstrumentUpdate
    ) -> Instrument:
        """Update equipment"""
        equipment = EquipmentService.get_equipment_by_id(db, equipment_id)
        
        update_data = updates.dict(exclude_unset=True)
        
        # Check serial number uniqueness if being updated
        if "serial_number" in update_data:
            existing = db.query(Instrument).filter(
                Instrument.serial_number == update_data["serial_number"],
                Instrument.id != equipment_id
            ).first()
            if existing:
                raise HTTPException(
                    status_code=400,
                    detail="Equipment with this serial number already exists"
                )

        for field, value in update_data.items():
            setattr(equipment, field, value)

        db.commit()
        db.refresh(equipment)
        return equipment

    @staticmethod
    def delete_equipment(db: Session, equipment_id: int) -> None:
        """Delete equipment"""
        equipment = EquipmentService.get_equipment_by_id(db, equipment_id)
        db.delete(equipment)
        db.commit()

    @staticmethod
    def add_calibration(
        db: Session,
        equipment_id: int,
        calibration: CalibrationCreate
    ) -> InstrumentCalibration:
        """Add calibration record"""
        equipment = EquipmentService.get_equipment_by_id(db, equipment_id)
        
        db_calibration = InstrumentCalibration(
            **calibration.dict(),
            instrument_id=equipment_id,
            calibration_date=datetime.utcnow()
        )
        
        db.add(db_calibration)
        EquipmentService.check_calibration_status(equipment)
        db.commit()
        db.refresh(db_calibration)
        return db_calibration

    @staticmethod
    def add_maintenance(
        db: Session,
        equipment_id: int,
        maintenance: MaintenanceCreate
    ) -> InstrumentMaintenanceLog:
        """Add maintenance record"""
        equipment = EquipmentService.get_equipment_by_id(db, equipment_id)
        
        db_maintenance = InstrumentMaintenanceLog(
            **maintenance.dict(),
            instrument_id=equipment_id,
            maintenance_date=datetime.utcnow()
        )
        
        db.add(db_maintenance)
        db.commit()
        db.refresh(db_maintenance)
        return db_maintenance

    @staticmethod
    async def add_attachment(
        db: Session,
        equipment_id: int,
        file: UploadFile
    ) -> Dict[str, Any]:
        """Add attachment to equipment"""
        equipment = EquipmentService.get_equipment_by_id(db, equipment_id)
        
        # Save file and get URL
        file_url = await save_upload_file(
            file,
            folder=f"equipment/{equipment_id}/attachments"
        )
        
        attachment_data = {
            "name": file.filename,
            "type": file.content_type,
            "url": file_url,
            "uploaded_at": datetime.utcnow()
        }
        
        return attachment_data

    @staticmethod
    def get_equipment_types() -> List[Dict[str, Any]]:
        """Get all equipment types"""
        return [
            {"id": i, "value": t.value, "description": t.value}
            for i, t in enumerate(EquipmentType, 1)
        ]

    @staticmethod
    def get_equipment_statuses() -> List[Dict[str, Any]]:
        """Get all equipment statuses"""
        return [
            {"id": i, "value": s.value, "description": s.value}
            for i, s in enumerate(EquipmentStatus, 1)
        ]

    @staticmethod
    def add_note(
        db: Session,
        equipment_id: int,
        note: NoteCreate
    ) -> Dict[str, Any]:
        """Add note to equipment"""
        equipment = EquipmentService.get_equipment_by_id(db, equipment_id)
        
        db_note = Note(
            content=note.content,
            user=note.user,
            instrument_id=equipment_id,
            timestamp=datetime.utcnow()
        )
        
        db.add(db_note)
        db.commit()
        db.refresh(db_note)
        return db_note

    @staticmethod
    def check_calibration_status(equipment: Instrument) -> None:
        """Update equipment calibration status"""
        if not equipment.calibrations:
            return
        
        latest = max(equipment.calibrations, key=lambda c: c.calibration_date)
        if datetime.utcnow() > latest.due_date:
            equipment.status = EquipmentStatus.OUT_OF_SERVICE
